src/core/lib/gpr/arena.cc

```diff
@@ -120,51 +120,56 @@ size_t gpr_arena_destroy(gpr_arena* arena) {
 
 void* gpr_arena_alloc(gpr_arena* arena, size_t size) {
   size = GPR_ROUND_UP_TO_ALIGNMENT_SIZE(size);
-  size_t start = static_cast<size_t>(
+  size_t previous_size_of_arena_allocations = static_cast<size_t>(
       gpr_atm_no_barrier_fetch_add(&arena->size_so_far, size));
-  size_t total_size_of_arena_allocations = start + size;
+  size_t updated_size_of_arena_allocations =
+      previous_size_of_arena_allocations + size;
   zone* z = &arena->initial_zone;
-
-  // Check to see if the allocation isn't able to start in the initial zone.
+  // Check to see if the allocation isn't able to end in the initial zone.
   // This statement is true only in the uncommon case because of our arena
   // sizing historesis (that is, most calls should have a large enough initial
   // zone and will not need to grow the arena).
-  if (start >= z->size_end) {
-    // Find the correct zone to start the allocation in
+  if (updated_size_of_arena_allocations > z->size_end) {
+    // Find a zone to fit this allocation
     gpr_mu_lock(&arena->arena_growth_mutex);
-    while (start >= z->size_end) {
+    while (updated_size_of_arena_allocations > z->size_end) {
       if (z->next == nullptr) {
-        size_t next_z_size = total_size_of_arena_allocations;
+        size_t next_z_size = updated_size_of_arena_allocations;
         z->next = static_cast<zone*>(zalloc_aligned(
             GPR_ROUND_UP_TO_ALIGNMENT_SIZE(sizeof(zone)) + next_z_size));
         z->next->size_begin = z->size_end;
         z->next->size_end = z->size_end + next_z_size;
+        // Note that we do an extra increment of size_so_far to prevent multiple
+        // simultaneous callers from stepping on each other. However, this extra
+        // increment means some space in the arena is wasted.
+        // So whenever we need to allocate x bytes and there are x - n (where
+        // n > 0) remaining in the current zone, we will waste x bytes (x - n
+        // in the current zone and n in the new zone).
+        previous_size_of_arena_allocations = static_cast<size_t>(
+            gpr_atm_no_barrier_fetch_add(&arena->size_so_far, size));
+        updated_size_of_arena_allocations =
+            previous_size_of_arena_allocations + size;
       }
       z = z->next;
     }
     gpr_mu_unlock(&arena->arena_growth_mutex);
   }
-  // Now that we have a zone we could start in, can the allocation fit in this
-  // zone? If not, call arena_alloc again. Note that this means some space in
-  // the arena is wasted because we have already incremented the size_so_far of
-  // the arena and we'll increment it again for the same attempted allocation.
-  if (total_size_of_arena_allocations > z->size_end) {
-    return gpr_arena_alloc(arena, size);
-  }
-  GPR_ASSERT(start >= z->size_begin);
-  GPR_ASSERT(total_size_of_arena_allocations <= z->size_end);
+
```

Comment:
Nit: Please eliminate unnecessary blank lines within functions.

Reply:
I'll come talk to you offline about this. I think the use of blank lines here within the function is consistent with Google's style guide, but I would like to hear more of your reasoning / opinion here :)